### ISSUE TITLE
Feature: Parallel outbound sync

### ISSUE THREAD

**Issue by touilleMan** [Scille/parsec-cloud#7760]
# Parallel outbound sync

## Current situation

Currently the outbound sync system doesn't do any parallel operation:

- Outbound sync monitor only trigger one entry synchronization at a time (https://github.com/Scille/parsec-cloud/blob/abd201a0301451774bc1ed4764f112bcb7ff4d4a/libparsec/crates/client/src/monitors/workspace_outbound_sync.rs#L120-L152)
- `WorkspaceOps::outbound_sync()` also (in the case of a file) uploads each block one by one (https://github.com/Scille/parsec-cloud/blob/abd201a0301451774bc1ed4764f112bcb7ff4d4a/libparsec/crates/client/src/workspace/transactions/outbound_sync.rs#L757-L819)

This works, but needlessly slow down the synchonization process (typically a lot of time is lost waiting for server reply).

## Non-goal

Here we only focus on the outbound sync (i.e. the upload of new data on the server).

It's possible we may want to be able to do parallel download of data (typically when
implementing the remanence, where enabling remanence on a workspace trigger a full
download of all current data). However this is a topic for another time.

## Synchronization scenarios

To choose the right strategy for parallele synchronization, we should first consider the scenarios our system will have to face.
We can roughly consider 3 main different scenarios:

a) One big file (say 1000 blocks of 512Ko => 512Mo)
b) Multiple small files (say 1000 folders)
c) A mix of the two (say 10 files of 200x512Ko, 100 files of 1x512Ko, 100 folders)

Case c) is insteresting in that it showcases nices properties other than raw throughput:

- Big files shouldn't take the whole bandwidth: small modifications are the most common and are fast
  to upload, so it's good to always keep some space for them.
- It's better to completly synchronize one file than to have 10 in progress taking 10x more time to finish.

## Possible architecture

The idea revolve around a new structure: `UploadBandwithLimiter`.

As it names implies, this structure keep track of what is currently under upload
and is in charge of deciding if new upload jobs should be started.

Notes:

- There is multiple reason to send request to the server, however most of them upload
  a negligeable quantity of data (e.g. a manifest or a certificate which are < 1Ko). So
  in the end we only care about the upload of blocks (and we don't want anyway to slow
  down feature such as worskpace sharing because a synchronization is under way !).
- This `UploadBandwithLimiter` structure doesn't need to be related to the
  `AuthenticatedCmds` structure that is used to actually send request to the server.
  Actually it is much better to keep both totally separated as this simplify the code (
  also see previous point about most query not needing bandwith control).
- In theory, the bandwith limit should be global to the program (since the network
  bandwith *is* something globally shared on the machine). However this brings complexity
  (since each libparsec client is isolated) for few benefit (most of the time a single
  libparsec client is started, and if multiple clients are started it's likely a single
  one is doing sync work anyway). So it's just simpler to keep the bandwith limiter at
  the client level.

Hence the bandwith limiter is a simple class that is shared between the `Client` and the
`WorkspaceOps` instances as an `Arc<UploadBandwithLimiter>`.

The basic idea is:

- In the outbound sync monitor, when [Action::DueTimeReached](https://github.com/Scille/parsec-cloud/blob/abd201a0301451774bc1ed4764f112bcb7ff4d4a/libparsec/crates/client/src/monitors/workspace_outbound_sync.rs#L217)
  is handled, a new coroutine is spawned and runs `WorkspacesOps::outbound_sync`. This also increase a counter
  of currently executing sync jobs (each job should also be checked to be still running [in the monitor loop](https://github.com/Scille/parsec-cloud/blob/abd201a0301451774bc1ed4764f112bcb7ff4d4a/libparsec/crates/client/src/monitors/workspace_outbound_sync.rs#L185)).
- There is limit on the number of job that can run at a time (let say 10). Once it is reached, the
  `Action::DueTimeReached` handler only increase the due time (typically to another second) and go
  back to sleep.
- In `WorkspaceOps::outbound_sync`, first the bandwith limiter is called to register an upload
  job (this is done by passing the total number of blocks that are to be uploaded, this way
  the bandwith limiter knows if we want to upload a big file or something smaller that should
  have more priority). This call returns a structure that represent the parallelism that is
  allowed for uploading data, and has a destructor to inform the bandwith limiter when the
  reserved resources are no longer needed.
- If there is already too many upload jobs registered, `WorkspaceOps::outbound_sync` returns a special
  `OutboundSyncOutcome::NoMoreUploadBandwith` status. In this case, the outbound sync monitor should consider it's
  not the right time to sync this file and re-enqueue it in it own need-to-be-sync queue (this is very similar to what is done for [OutboundSyncOutcome::EntryIsBusy](https://github.com/Scille/parsec-cloud/blob/abd201a0301451774bc1ed4764f112bcb7ff4d4a/libparsec/crates/client/src/workspace/transactions/outbound_sync.rs#L68))
- In `WorkspaceOps::outbound_sync` (well in [`upload_blocks` to be more precise](https://github.com/Scille/parsec-cloud/blob/abd201a0301451774bc1ed4764f112bcb7ff4d4a/libparsec/crates/client/src/workspace/transactions/outbound_sync.rs#L753)), as many
  coroutines are spawned as it is allowed. Those coroutines are used to upload the blocks in parallel.

Note the limit in the outbound sync monitor, this is required given we don't want to expose
the bandwith limiter to it (it would complexify the code, especially given atomicity has to be accounted for).
Instead we have a simple way to know "there is too much going on for now" to avoid back pressure.

Similarly, the `OutboundSyncOutcome::NoMoreUploadBandwith` is useful to avoid having multiple big file
taking all the coroutines in the monitor (and being idle due to the bandwith limiter) and creating a
famine for smaller files.

Regarding the strategy to allocate bandwith the limiter could use:

- Split files into 2 buckets: small (0 or 1 blocks, so folders are always small), big ( > 1 blocks)
- Always allow up to 4 small blocks, no matter how many big are allocated.
- For small, don't allow blocks upload paralellism (there is at most 1 block to upload !)
- For big, allow 8 blocks upload parallelism. And no parallel jobs (we sould focus on uploading one big file at a time).
- For small, allow 10 parallelism.

So for instance:

- 12 new small files are created, the first 10 start upload
- 2 new big files are added, one start upload
- no new file start upload until small files goes under 4 (in which case another
  small starts), or the first big is done (in which case the other big starts)
